### ISSUE TITLE
[2019 ICLR] An Empirical Study of Example Forgetting during Deep Neural Network Learning

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#76]
Classification task에서 neural network를 학습하면 example forgetting이라는 현상이 발생하며, 이를 분석한 논문

논문의 핵심만 간단하게 요약

## 1. Intuition

Continual learning 분야에서 neural network가 catastrophic forgetting이라는 현상을 겪음

Catastrophic forgetting이란 neural network가 새로운 task를 학습할 때, 전에 배웠던 정보를 잊는 것을 의미함

저자들은 동일한 task에서도 해당 현상이 일어나는지에 대한 실험을 진행하며, (catastrophic) example forgetting이라는 현상을 관측함
(each mini-batch ≈ mini-task)

## 2. Defining and Computing Example Forgetting

Event의 대상은 1개의 data point로, 이에 대한 acc를 측정하면 0 혹은 1인 binary 값을 가짐

- Forgetting event : acc_t > acc_t+1
- Learning event : acc_t < acc_t+1
- Unforgettable example : example that learn at some point and experience no forgetting event during the whole course of training

### 3.1. Computing Forgetting Examples
<img width="1164" alt="image" src="https://user-images.githubusercontent.com/49643709/173853258-24a1be8d-8c17-40bd-9724-847a198562f8.png">

모든 step마다 모든 data point에 acc를 구하여 acc log를 구하면 시간이 너무 오래 걸림

이를 단순화한 `Algorithm 1`을 이용하여 mini-batch가 주어졌을 때, 해당 gradient update 전에 mini-batch data point의 acc만 측정하여 acc log를 저장

### 3.2. Number of Forgetting Events
<img width="1100" alt="image" src="https://user-images.githubusercontent.com/49643709/173855827-816d804c-4f61-44fe-9472-e2d8c98a9130.png">

Acc log를 기반으로 data point가 forgetting 현상이 일어난 횟수를 count하여 dataset별로 histogram plot

### 3.3. Stability across Seeds
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/49643709/173856764-923bca0d-470c-4c6c-9f47-80af15b730b2.png">

단순화한 `Algorithm 1`을 이용하여 1개의 seed에 대해서 측정을 하면 여러 random성으로 인해 variance가 클 수 있음

저자들은 이를 위해 5개의 random seed에 대해 구한 acc log를 average함

이렇게 average acc log가 stable하다는 것을 보이기 위해 100개의 random seed에 대한 acc log를 구한 후, 5개씩 묶어 20개의 average acc log를 구함

이에 대한 mean과 confidence를 구하여 plot한 결과가 `Figure 14`이며, forgetting event가 거의 일어나지 않은 examples에 대해서 confident하기에, average acc log를 사용해도 괜찮음을 보임

### 3.4. Forgetting by Chance
<img width="969" alt="image" src="https://user-images.githubusercontent.com/49643709/174008438-172e148a-7302-4b99-91ff-e3fb0536029d.png">
<img width="991" alt="image" src="https://user-images.githubusercontent.com/49643709/174227711-b1c3a42e-e6e5-4308-9c88-0cf2eaad4ba0.png">

Forgetting event가 다른 mini-batch를 배우면서 잊는 것이 아니라 random gradient가 발생해도 일어날 수 있는데, 이러한 경우를 forgetting event가 `chance`로 발생했다고 표현함

그럼 어떻게 random gradient를 구할 수 있는지에 대해서는 다음과 같음

1. Base network는 일반적인 학습을 하고 gradient를 저장함
2. Base network와 동일한 initial weight를 가지고 있는 clone network를 생성
3. 1번에서 구한 gradient를 random shuffle하여 clone network update

<img width="531" alt="image" src="https://user-images.githubusercontent.com/49643709/174227552-30e54247-266b-434c-b014-35e576f891d4.png">

이렇게 하면 최종 weight는 같은데 weight가 optimize되는 방향이 random임
(weight vector의 차원이 고차원이며, 또한 iteration이 많기에 이를 mixing하면 충분히 random하다고 볼 수 있음)

Random gradient를 통해 update를 했을 때 acc log를 이용하여 forgetting event를 측정하여 plot한 결과, forget event가 일어나는 횟수는 최대 2를 넘지 않으며, 대부분 1 미만으로 일어남을 확인할 수 있음

이를 통해 우연히 forgetting event가 일어났다는 것은 무시해도 된다고 볼 수 있음

### 3.5. First Learning Events
<img width="968" alt="image" src="https://user-images.githubusercontent.com/49643709/174227935-2af0c09e-e000-4642-ad9f-54d47e6387c4.png">

First learning event가 일어나는 시점은 unforgettable examples, forgettable examples 대부분이 학습 초기에 일어남

다만 forgettable examples의 경우, first learning event가 일어나는 시점이 unforgettable event보다 상대적으로 늦음

### 3.6. Misclassification Margin
<img width="1047" alt="image" src="https://user-images.githubusercontent.com/49643709/173870433-bd6c7a06-beca-45d8-a292-290b8f153fc0.png">
<img width="972" alt="image" src="https://user-images.githubusercontent.com/49643709/174229797-fb389f6a-b4f3-4470-bf56-19a8633b0789.png">

저자들은 classification margin이라는 metric을 data point의 true label의 logit과 other labels 중 max logit을 뺀 값으로 정의함

Classification margin이라는 metric을 이용하여 misclassified image에 대하여 얼마나 confident를 가지고 틀리는지를 확인할 수 있음

Forgetting event가 자주 일어난 frequently forgettable examples의 경우, 확신을 가지고 틀린다는 것을 확인할 수 있음

### 3.7. Visual Inspection
<img width="1215" alt="image" src="https://user-images.githubusercontent.com/49643709/173871251-c7257e86-7182-4ad4-a771-39c9fa430970.png">

그렇다면 frequently forgettable examples와 unforgettable examples가 뭔지 이에 해당하는 이미지들을 시각화함

Unforgettable examples는 상대적으로 물체가 중앙에 위치해있으며, 해당 label의 특징이 전반적으로 포함되어 있음

그에 반해 frequently forgettable examples는 다른 label에서 볼 수 있을만한 atypical class characteristic을 가짐을 확인할 수 있음

### 3.8. Detecting of Noisy Examples
<img width="888" alt="image" src="https://user-images.githubusercontent.com/49643709/173871914-0e2d8ce0-5ea3-4fbd-8539-ddf0dc9f1f98.png">

Noisy-labeled examples의 경우, 해당 label이 noisy하기에 atypical class characteristic을 가짐

Noisy-labeled dataset을 학습하여 acc log를 이용하여 plot한 결과, noisy-labeled data의 경우 무조건 forgetting event가 발생하며 regular examples과 다른 분포를 가짐을 확인할 수 있음
(`Figure 3`의 왼쪽 graph)

그렇다면 noisy-labeled dataset의 이미지가 원래 forgetting event가 발생하는 데이터가 아닌지에 대해 noisy-labeled dataset image에 대해 원래 label을 붙인 경우와 noisy-label을 붙인 경우의 histogram을 비교하여 noisy-label을 붙였기에 발행한 것임을 보임

### 3.9. Continual Learning Setup
<img width="1014" alt="image" src="https://user-images.githubusercontent.com/49643709/173872828-abb214b7-4aa7-482b-b3fc-61a497b65899.png">

저자들은 실험을 통해 같은 동일한 task에서 example forgetting 현상이 일어남을 확인하였으니, 원래의 intuition인 catastrophic forgetting이 발생하는 continual learning과 비슷하게 셋업하여 실험을 진행

Random하게 dataset을 나눈 경우와 forgettable/unforgettable dataset으로 나눈 경우 모두 catastrophic forgetting 현상은 일어남

다만 forgettable/unforgettable dataset으로 나눈 경우 random dataset과 다른 특성이 나타남

- Unforgettable dataset을 학습하게 되면 forgettable dataset의 acc가 많이 떨어짐
- Forgettable dataset을 학습하게 되면 unforgettable dataset의 acc는 적게 떨어짐

이를 통해 unforgettable examples는 forgettable examples에 비해 상대적으로 information의 양이 적다는 것을 유추할 수 있음

### 4.1. Fraction of Training Dataset Generalization Performance
<img width="964" alt="image" src="https://user-images.githubusercontent.com/49643709/174232995-84e9ad34-c9ce-4476-be5e-70ba86e07cdf.png">

Unforgettable example의 경우 정보양이 적음을 유추할 수 있는데 이를 실험을 통해 더욱 구체화함

Random remove 했을 때의 generalization performance와 forgetting event가 적게 일어난 순서대로 remove한 경우의 generalization performance를 비교하여, unforgettable examples는 remove해도 generalization performance에 영향을 주지 않는다는 것을 보여줌

<img width="959" alt="image" src="https://user-images.githubusercontent.com/49643709/174233026-579dc77b-974b-4b42-83d0-50f5d632cb20.png">

저자들은 MNIST, permutedMNIST, CIFAR-10 dataset에 대해 forgetting event가 적게 일어난 순서대로 remove하였을 때 generalization performance를 plot

저자들은 MNIST, permutedMNIST, CIFAR-10 dataset에 대해 5개의 random seed에서 common하게 일어난 unforgettable example의 비율을 구했는데 이는 다음과 같음

`MNIST (91.7%) > permutedMNIST (75.3%) > CIFAR-10 (31.3%)`

`Figure 6`의 오른쪽 그래프를 보면, test error가 증가하는 시점이 unforgettable examples의 비율일 때임을 확인할 수 있음

즉 unforgettable examples는 제거해도 generalization performance에는 영향을 주지 않는다는 의미임

### 4.2. Support Vectors

Gradient descent로 학습한 linear network가 maximum margin classifier가 되게 학습한다는 recent work가 있음

Maximum margin classifier가 decision boundary를 학습하는데 있어 data마다 중요도가 다름
(어떤 data point는 다른 data point에 비해 중요하지 않음)

Dataset의 intrinsic dimension이 높을수록 support vector의 수 또한 많아지며, 이에 따른 필요로 하는 중요한 data point가 많아지게 됨

Dataset의 data들을 알기에, intrinsic dimension은 MNIST < permutedMNIST < CIFAR-10임을 알 수 있으며 이에 따른 support vector와 중요한 data point의 개수 또한 많아짐을 알 수 있음

즉 해당 관점에서 봤을 때, 5개의 random seed에서 common하게 일어난 unforgettable example의 비율은 `MNIST (91.7%) > permutedMNIST (75.3%) > CIFAR-10 (31.3%)`이라는 결과와 일치함
(dataset의 intrinsic dimension이 높을수록 support vector의 수가 많아지기에, 중요하지 않은 data의 수가 적어진다는 의미는 unforgettable example의 수가 적어진다는 말)

## 5. Transferable Forgetting Events
<img width="1138" alt="image" src="https://user-images.githubusercontent.com/49643709/173876619-69e9e5da-d76b-4490-8974-9e7fdc6ef4cc.png">

200 epoch나 75 epoch나 data point의 forgetting events는 거의 똑같음

또한 가벼운 네트워크로 학습한 acc log와 무거운 네트워크로 학습한 acc log가 비슷함

가벼운 네트워크에서의 acc log를 통해 구한 unforgettable examples를 빼가면서 무거운 네트워크를 학습해도 generalization performance가 안떨어짐

해당 실험 결과를 통해 2가지를 알 수 있음

1. Unforgettable examples는 transfer가 가능함
2. 가벼운 네트워크에서 적당한 epoch로 학습하여 구한 acc log를 이용하여 unforgettable examples를 구하면 computing 시간을 크게 줄일 수 있으면서 비슷한 효과를 낼 수 있음

**Comment by Jasonlee1995** [comment]
[example_forgetting.pdf](https://github.com/Jasonlee1995/AI_Papers/files/8911302/example_forgetting.pdf)
